### PR TITLE
Implement language-agnostic proto tests

### PR DESCRIPTION
**Pull request by jskeet**
Definitely review one commit at a time.

The second commit mostly changes how we handle sentinel values, including making sure that we optimize away unnecessary updates.

Having implemented these tests, I've given feedback which will probably create *more* tests and require more changes - but I wanted to get these in first.

### CODE REVIEW

File: apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data.Tests/Proto/ProtoTest.cs [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+                default:
+                    throw new ArgumentException($"Unknown value kind: {value.KindCase}");
+            }
+        }
+
+        // Rather than using Moq to expect a commit, we just have a small recording client. This makes
+        // it easier to get good diagnostics about what went wrong.
+        private class MockCommitClient : FirestoreClient
+        {
+            private readonly CommitRequest m_expectedRequest;
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data.Tests/Proto/ProtoTest.cs]: Why the "m" prefix?

**Comment by chrisdunelm** [comment]
Oops, hang on. I LGTM'ed after only looking at the first commit.

File: apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data.Tests/Proto/ProtoTest.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data.Tests/Proto/ProtoTest.cs]: Whoops, had forgotten our convention. (I've used too many conventions in too many different places...)

File: apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/WriteBatch.cs [hunk 2]
```diff
@@ -58,16 +57,13 @@ public WriteBatch Create(DocumentReference documentReference, object documentDat
...
-                {
-                    Fields = { ValueSerializer.SerializeMap(documentData) },
-                    Name = documentReference.Path
-                }
-            };
-            AddUpdate(write);
+            var fields = ValueSerializer.SerializeMap(documentData);
+            var serverTimestamps = new List<FieldPath>();
+            var deletes = new List<FieldPath>();
+            FindSentinels(fields, FieldPath.Empty, serverTimestamps, deletes);
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/WriteBatch.cs]: Could make `FindSentinels` return a `(List<FieldPath>, List<FieldPath>)` tuple?
(Or even better, use `ImmutableList<FieldPath>`)

Comment (jskeet) [reply apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/WriteBatch.cs]: I suspect we'll want to introduce a dependency on ValueTuple eventually, but I'd rather put that off for now. (Especially as I seem to remember there are some issues with ValueTuple dependencies due to a mistake in one version. I'd need to validate that though.)

Fortunately all of this is internal - we can do it later if we want.

**Comment by jskeet** [comment]
Will fix the m_, rebase and merge on green.

**Pull request merged**